### DISCUSSION TITLE
Is this a bug? graphql.service.ts::buildFilterQuery fails when using directly using a complex object

### DISCUSSION THREAD

**Discussion by Harsgalt86** [ghiscoding/slickgrid-universal#1128] [Q&A] [answered]
When the column has a complex object and references it's nested fields, a null pointer exception occurs. 
eg:  These having a direct reference to "billing" in the array ['billing', 'billing.address.street', 'billing.address.zip'] causes a crash

Can reproduce by adding this test in `graphql.service.spec.ts`
```
it('should return a query string including a direct reference to a complex object', () => {
  const expectation = `firstName, lastName, billing{address{street, zip}}`;
  const columns = ['firstName', 'lastName', 'billing', 'billing.address.street', 'billing.address.zip'];

  const query = service.buildFilterQuery(columns);

  expect(removeSpaces(query)).toBe(removeSpaces(expectation));
});
```

I've managed a fix by changing the `set` function in `graphql.service.ts::buildFilterQuery` to
```
const set = (o: any = {}, a: any) => {
      const k = a.shift();
      o[k] = a.length ? set(o[k] ?? undefined, a) : null;
      return o;
    };
```
By passing `undefined` it will use the default `{}` as per the parameter signature.

**Comment by ghiscoding** [comment]
I haven't used the GraphQL one in a while, please open a Pull Request if you found an issue and a fix. Thank you. 

Also FYI, the [Example 10](https://ghiscoding.github.io/slickgrid-universal/#/example10) does use complex object as can be seen in this column, that demo only shows the query string since I didn't have a GraphQL server to show the demo, but typically the query string is enough to demo the feature. 

https://github.com/ghiscoding/slickgrid-universal/blob/0d10202895fb958a5cd3a5fe530c4eeea931fd1c/examples/vite-demo-vanilla-bundle/src/examples/example10.ts#L106-L108

isn't your proposed code doing the same thing though? I mean `??` is already for `undefined` (or `null` as per [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing)) so it seems like you're saying if `set(o[k]` is `undefined` then pass `undefined`... so why would we do that? 

Just for reference, I assume you're looking at this piece of code, I would be happy to receive any PR with an explanation. thanks

https://github.com/ghiscoding/slickgrid-universal/blob/37d0c3c7f48b8c0b87d2542e955a6ec330abf939/packages/graphql/src/services/graphql.service.ts#L196-L200

**Reply by Harsgalt86** [reply, marked as answer]
Thanks, I've opened a PR request with an explanation of why "null coalescing" to `undefined` works
https://github.com/ghiscoding/slickgrid-universal/pull/1130

but will also address here
set(o[k] is undefined then pass undefined // this is true, but I'm also saying
set(o[k] is null then pass undefined // which makes use of the property's default property which null doesn't (null stays null)
